[3.10.1] fails to build with poppler-25.02.0

### What is the bug?

3.10.1 fails to build with poppler-25.02.0:

>[ 41%] Building CXX object frmts/pcidsk/sdk/CMakeFiles/pcidsk_sdk.dir/segment/cpcidsk_tex.cpp.o
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp: In member function ‘CPLErr PDFDataset::ReadPixels(int, int, int, int, GSpacing, GSpacing, GSpacing, GByte*)’:
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:1978:38: error: cannot convert ‘std::unique_ptr<OCGs>’ to ‘OCGs*’ in initialization
   1978 |         OCGs *poOldOCGs = poCatalog->optContent;
        |                           ~~~~~~~~~~~^~~~~~~~~~
        |                                      |
        |                                      std::unique_ptr<OCGs>
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:1992:37: error: no match for ‘operator=’ (operand types are ‘std::unique_ptr<OCGs>’ and ‘OCGs*’)
   1992 |             poCatalog->optContent = poOldOCGs;
        |                                     ^~~~~~~~~
  In file included from /usr/include/c++/14.2.0/memory:78,
                   from /usr/include/poppler/goo/GooString.h:45,
                   from /usr/include/poppler/Object.h:45:
  /usr/include/c++/14.2.0/bits/unique_ptr.h:424:9: note: candidate: ‘template<class _Up, class _Ep> typename std::enable_if<std::__and_<std::__and_<std::is_convertible<typename std::unique_ptr<_Up, _Ep>::pointer, typename std::__uniq_ptr_impl<_Tp, _Dp>::pointer>, std::__not_<std::is_array<_Up> > >, std::is_assignable<_Dp&, _Ep&&> >::value, std::unique_ptr<_Tp, _Dp>&>::type std::unique_ptr<_Tp, _Dp>::operator=(std::unique_ptr<_Up, _Ep>&&) [with _Ep = _Up; _Tp = OCGs; _Dp = std::default_delete<OCGs>]’
    424 |         operator=(unique_ptr<_Up, _Ep>&& __u) noexcept
        |         ^~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:424:9: note:   template argument deduction/substitution failed:
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:1992:37: note:   mismatched types ‘std::unique_ptr<_Tp, _Dp>’ and ‘OCGs*’
   1992 |             poCatalog->optContent = poOldOCGs;
        |                                     ^~~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:408:19: note: candidate: ‘std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(std::unique_ptr<_Tp, _Dp>&&) [with _Tp = OCGs; _Dp = std::default_delete<OCGs>]’
    408 |       unique_ptr& operator=(unique_ptr&&) = default;
        |                   ^~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:408:29: note:   no known conversion for argument 1 from ‘OCGs*’ to ‘std::unique_ptr<OCGs>&&’
    408 |       unique_ptr& operator=(unique_ptr&&) = default;
        |                             ^~~~~~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:434:7: note: candidate: ‘std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(std::nullptr_t) [with _Tp = OCGs; _Dp = std::default_delete<OCGs>; std::nullptr_t = std::nullptr_t]
    434 |       operator=(nullptr_t) noexcept
        |       ^~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:434:17: note:   no known conversion for argument 1 from ‘OCGs*’ to ‘std::nullptr_t’
    434 |       operator=(nullptr_t) noexcept
        |                 ^~~~~~~~~
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:1998:33: error: no match for ‘operator=’ (operand types are ‘std::unique_ptr<OCGs>’ and ‘OCGs*’)
   1998 |         poCatalog->optContent = poOldOCGs;
        |                                 ^~~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:424:9: note: candidate: ‘template<class _Up, class _Ep> typename std::enable_if<std::__and_<std::__and_<std::is_convertible<typename std::unique_ptr<_Up, _Ep>::pointer, typename std::__uniq_ptr_impl<_Tp, _Dp>::pointer>, std::__not_<std::is_array<_Up> > >, std::is_assignable<_Dp&, _Ep&&> >::value, std::unique_ptr<_Tp, _Dp>&>::type std::unique_ptr<_Tp, _Dp>::operator=(std::unique_ptr<_Up, _Ep>&&) [with _Ep = _Up; _Tp = OCGs; _Dp = std::default_delete<OCGs>]’
    424 |         operator=(unique_ptr<_Up, _Ep>&& __u) noexcept
        |         ^~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:424:9: note:   template argument deduction/substitution failed:
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:1998:33: note:   mismatched types ‘std::unique_ptr<_Tp, _Dp>’ and ‘OCGs*’
   1998 |         poCatalog->optContent = poOldOCGs;
        |                                 ^~~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:408:19: note: candidate: ‘std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(std::unique_ptr<_Tp, _Dp>&&) [with _Tp = OCGs; _Dp = std::default_delete<OCGs>]’
    408 |       unique_ptr& operator=(unique_ptr&&) = default;
        |                   ^~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:408:29: note:   no known conversion for argument 1 from ‘OCGs*’ to ‘std::unique_ptr<OCGs>&&’
    408 |       unique_ptr& operator=(unique_ptr&&) = default;
        |                             ^~~~~~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:434:7: note: candidate: ‘std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(std::nullptr_t) [with _Tp = OCGs; _Dp = std::default_delete<OCGs>; std::nullptr_t = std::nullptr_t]’
    434 |       operator=(nullptr_t) noexcept
        |       ^~~~~~~~
  /usr/include/c++/14.2.0/bits/unique_ptr.h:434:17: note:   no known conversion for argument 1 from ‘OCGs*’ to ‘std::nullptr_t’
    434 |       operator=(nullptr_t) noexcept
        |                 ^~~~~~~~~
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp: In member function ‘void PDFDataset::ExploreLayersPoppler(GDALPDFArray*, int, int, CPLString, int, int&, bool&)’:
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:3739:77: error: invalid conversion from ‘const OCGs*’ to ‘OCGs*’ [-fpermissive]
   3739 |                 OCGs *optContentConfig = m_poDocPoppler->getOptContentConfig();
        |                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
        |                                                                             |
        |                                                                             const OCGs*
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp: In member function ‘void PDFDataset::FindLayersPoppler(int)’:
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:3775:65: error: invalid conversion from ‘const OCGs*’ to ‘OCGs*’ [-fpermissive]
   3775 |     OCGs *optContentConfig = m_poDocPoppler->getOptContentConfig();
        |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
        |                                                                 |
        |                                                                 const OCGs*
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:3779:51: error: invalid conversion from ‘const Array*’ to ‘Array*’ [-fpermissive]
   3779 |     Array *array = optContentConfig->getOrderArray();
        |                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
        |                                                   |
        |                                                   const Array*
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp: In member function ‘void PDFDataset::TurnLayersOnOffPoppler()’:
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:3815:65: error: invalid conversion from ‘const OCGs*’ to ‘OCGs*’ [-fpermissive]
   3815 |     OCGs *optContentConfig = m_poDocPoppler->getOptContentConfig();
        |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
        |                                                                 |
        |                                                                 const OCGs*
  [ 41%] Building CXX object frmts/pdf/CMakeFiles/gdal_PDF.dir/pdfio.cpp.o
  [ 41%] Building CXX object frmts/pcidsk/sdk/CMakeFiles/pcidsk_sdk.dir/segment/cpcidskbinarysegment.cpp.o
  make[2]: *** [frmts/pdf/CMakeFiles/gdal_PDF.dir/build.make:107: frmts/pdf/CMakeFiles/gdal_PDF.dir/pdfdataset.cpp.o] Error 1
  make[2]: *** Waiting for unfinished jobs....
  [ 41%] Building CXX object frmts/wcs/CMakeFiles/gdal_WCS.dir/gmlcoverage.cpp.o
  [ 41%] Building CXX object frmts/pcidsk/sdk/CMakeFiles/pcidsk_sdk.dir/segment/cpcidskbitmap.cpp.o
  [ 41%] Building CXX object frmts/wcs/CMakeFiles/gdal_WCS.dir/wcsdataset.cpp.o
  [ 41%] Building CXX object frmts/pcidsk/sdk/CMakeFiles/pcidsk_sdk.dir/segment/cpcidskblut.cpp.o
  [ 41%] Building CXX object frmts/http/CMakeFiles/gdal_HTTP.dir/httpdriver.cpp.o
  [ 41%] Building CXX object frmts/pcidsk/sdk/CMakeFiles/pcidsk_sdk.dir/segment/cpcidskpct.cpp.o
  [ 41%] Building CXX object frmts/pcidsk/sdk/CMakeFiles/pcidsk_sdk.dir/segment/cpcidskephemerissegment.cpp.o
  [ 41%] Built target gdal_HTTP
  [ 41%] Building CXX object frmts/wcs/CMakeFiles/gdal_WCS.dir/wcsdataset100.cpp.o
  [ 41%] Building CXX object frmts/pcidsk/sdk/CMakeFiles/pcidsk_sdk.dir/segment/cpcidskgcp2segment.cpp.o
  In file included from /home/SBo/gdal-3.10.1/frmts/pdf/pdfio.cpp:17:
  /home/SBo/gdal-3.10.1/frmts/pdf/pdfio.h:51:18: error: conflicting return type specified for ‘virtual void VSIPDFFileStream::reset()’
     51 |     virtual void reset() override;
        |                  ^~~~~
  In file included from /usr/include/poppler/Object.h:596,
                   from /home/SBo/gdal-3.10.1/frmts/pdf/pdfsdk_headers_poppler.h:45,
                   from /home/SBo/gdal-3.10.1/frmts/pdf/pdfsdk_headers_all.h:20,
                   from /home/SBo/gdal-3.10.1/frmts/pdf/gdal_pdf.h:39,
                   from /home/SBo/gdal-3.10.1/frmts/pdf/pdfio.cpp:13:
  /usr/include/poppler/Stream.h:129:18: note: overridden function is ‘virtual bool Stream::reset()’
    129 |     virtual bool reset() = 0;
        |                  ^~~~~
  make[2]: *** [frmts/pdf/CMakeFiles/gdal_PDF.dir/build.make:121: frmts/pdf/CMakeFiles/gdal_PDF.dir/pdfio.cpp.o] Error 1
  make[1]: *** [CMakeFiles/Makefile2:8080: frmts/pdf/CMakeFiles/gdal_PDF.dir/all] Error 2`

### Steps to reproduce the issue

see https://github.com/OSGeo/gdal/issues/11477

### Versions and provenance

Slackware Linux: slackware64-current (as of 5 Feb 2025)

### Additional context

_No response_

Hi, similar system to brobr, slackware64-current.

I've applied the last patch from above, 1b6d32d. Build still fails. Since the release tarball does not contain the autotest directory I downloaded the source code tarball that does and patch applies successfully. I get the same results with the release tarball if I comment out the autotest part. I can post the without patch output if needed. 

Here is the last few line of the output after patch has been applied:
```

[ 40%] Building CXX object frmts/pdf/CMakeFiles/gdal_PDF.dir/ogrpdflayer.cpp.o
[ 40%] Building CXX object frmts/pdf/CMakeFiles/gdal_PDF.dir/pdfcreatecopy.cpp.o
[ 40%] Building CXX object frmts/pdf/CMakeFiles/gdal_PDF.dir/pdfdataset.cpp.o
In file included from /home/non-slack/slackbuilds/gdal/tmp/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:31:
/home/non-slack/slackbuilds/gdal/tmp/gdal-3.10.1/frmts/pdf/pdfio.h:54:18: error: conflicting return type specified for ‘virtual bool VSIPDFFileStream::unfilteredReset()’
   54 |     virtual bool unfilteredReset() override;
      |                  ^~~~~~~~~~~~~~~
In file included from /usr/include/poppler/Object.h:596,
                 from /home/non-slack/slackbuilds/gdal/tmp/gdal-3.10.1/frmts/pdf/pdfsdk_headers_poppler.h:45,
                 from /home/non-slack/slackbuilds/gdal/tmp/gdal-3.10.1/frmts/pdf/pdfsdk_headers_all.h:20,
                 from /home/non-slack/slackbuilds/gdal/tmp/gdal-3.10.1/frmts/pdf/gdal_pdf.h:39,
                 from /home/non-slack/slackbuilds/gdal/tmp/gdal-3.10.1/frmts/pdf/pdfdataset.cpp:20:
/usr/include/poppler/Stream.h:216:18: note: overridden function is ‘virtual void Stream::unfilteredReset()’
  216 |     virtual void unfilteredReset() = 0;
      |                  ^~~~~~~~~~~~~~~
gmake[2]: *** [frmts/pdf/CMakeFiles/gdal_PDF.dir/build.make:107: frmts/pdf/CMakeFiles/gdal_PDF.dir/pdfdataset.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:9620: frmts/pdf/CMakeFiles/gdal_PDF.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

--- a/autotest/gdrivers/pdf.py
+++ b/autotest/gdrivers/pdf.py
@@ -1272,7 +1272,7 @@ def test_pdf_extra_rasters(poppler_or_pdfium):
             "LAYER_01_NAME=subbyte",
         ], "did not get expected layers"
     if pdf_is_poppler():
-        assert cs in (7926, 8177, 8174, 8165, 8172)
+        assert cs in (7926, 8177, 8174, 8165, 8172, 8193)
 
 
 ###############################################################################

--- a/frmts/pdf/pdfdataset.cpp
+++ b/frmts/pdf/pdfdataset.cpp
@@ -1964,6 +1964,14 @@ CPLErr PDFDataset::ReadPixels(int nReqXOff, int nReqYOff, int nReqXSize,
         PDFDoc *poDoc = m_poDocPoppler;
         poSplashOut->startDoc(poDoc);
 
+        // Note: Poppler 25.2 is certainly not the lowest version where we can
+        // avoid the hack.
+#if !(POPPLER_MAJOR_VERSION > 25 ||                                            \
+      (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION >= 2))
+#define USE_OPTCONTENT_HACK
+#endif
+
+#ifdef USE_OPTCONTENT_HACK
         /* EVIL: we modify a private member... */
         /* poppler (at least 0.12 and 0.14 versions) don't render correctly */
         /* some PDFs and display an error message 'Could not find a OCG with
@@ -1978,6 +1986,7 @@ CPLErr PDFDataset::ReadPixels(int nReqXOff, int nReqYOff, int nReqXSize,
         OCGs *poOldOCGs = poCatalog->optContent;
         if (!m_bUseOCG)
             poCatalog->optContent = nullptr;
+#endif
         try
         {
             poDoc->displayPageSlice(poSplashOut, m_iPage, m_dfDPI, m_dfDPI, 0,
@@ -1988,14 +1997,19 @@ CPLErr PDFDataset::ReadPixels(int nReqXOff, int nReqYOff, int nReqXSize,
         {
             CPLError(CE_Failure, CPLE_AppDefined,
                      "PDFDoc::displayPageSlice() failed with %s", e.what());
+
+#ifdef USE_OPTCONTENT_HACK
             /* Restore back */
             poCatalog->optContent = poOldOCGs;
+#endif
             delete poSplashOut;
             return CE_Failure;
         }
 
+#ifdef USE_OPTCONTENT_HACK
         /* Restore back */
         poCatalog->optContent = poOldOCGs;
+#endif
 
         SplashBitmap *poBitmap = poSplashOut->getBitmap();
         if (poBitmap->getWidth() != nReqXSize ||
@@ -3734,9 +3748,14 @@ void PDFDataset::ExploreLayersPoppler(GDALPDFArray *poArray,
                 }
                 else
                     osCurLayer = std::move(osName);
-                // CPLDebug("PDF", "Layer %s", osCurLayer.c_str());
+                    // CPLDebug("PDF", "Layer %s", osCurLayer.c_str());
 
-                OCGs *optContentConfig = m_poDocPoppler->getOptContentConfig();
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION >= 2)
+                const
+#endif
+                    OCGs *optContentConfig =
+                        m_poDocPoppler->getOptContentConfig();
                 struct Ref r;
                 r.num = poObj->GetRefNum().toInt();
                 r.gen = poObj->GetRefGen();
@@ -3772,11 +3791,19 @@ void PDFDataset::FindLayersPoppler(int iPageOfInterest)
     if (poPages)
         nPageCount = poPages->GetLength();
 
-    OCGs *optContentConfig = m_poDocPoppler->getOptContentConfig();
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION >= 2)
+    const
+#endif
+        OCGs *optContentConfig = m_poDocPoppler->getOptContentConfig();
     if (optContentConfig == nullptr || !optContentConfig->isOk())
         return;
 
-    Array *array = optContentConfig->getOrderArray();
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION >= 2)
+    const
+#endif
+        Array *array = optContentConfig->getOrderArray();
     if (array)
     {
         GDALPDFArray *poArray = GDALPDFCreateArray(array);
@@ -3812,7 +3839,11 @@ void PDFDataset::FindLayersPoppler(int iPageOfInterest)
 
 void PDFDataset::TurnLayersOnOffPoppler()
 {
-    OCGs *optContentConfig = m_poDocPoppler->getOptContentConfig();
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION >= 2)
+    const
+#endif
+        OCGs *optContentConfig = m_poDocPoppler->getOptContentConfig();
     if (optContentConfig == nullptr || !optContentConfig->isOk())
         return;
 

--- a/frmts/pdf/pdfio.cpp
+++ b/frmts/pdf/pdfio.cpp
@@ -237,23 +237,40 @@ int VSIPDFFileStream::lookChar()
 /*                                reset()                               */
 /************************************************************************/
 
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION >= 2)
+bool VSIPDFFileStream::reset()
+#else
 void VSIPDFFileStream::reset()
+#endif
 {
     nSavedPos = VSIFTellL(f);
     bHasSavedPos = TRUE;
     VSIFSeekL(f, nCurrentPos = nStart, SEEK_SET);
     nPosInBuffer = -1;
     nBufferLength = -1;
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION >= 2)
+    return true;
+#endif
 }
 
 /************************************************************************/
 /*                         unfilteredReset()                            */
 /************************************************************************/
 
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION > 2)
+bool VSIPDFFileStream::unfilteredReset()
+{
+    return reset();
+}
+#else
 void VSIPDFFileStream::unfilteredReset()
 {
     reset();
 }
+#endif
 
 /************************************************************************/
 /*                                close()                               */

--- a/frmts/pdf/pdfio.h
+++ b/frmts/pdf/pdfio.h
@@ -47,8 +47,20 @@ class VSIPDFFileStream final : public BaseStream
     virtual int getUnfilteredChar() override;
     virtual int lookChar() override;
 
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION >= 2)
+    virtual bool reset() override;
+#else
     virtual void reset() override;
+#endif
+
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION > 2)
+    virtual bool unfilteredReset() override;
+#else
     virtual void unfilteredReset() override;
+#endif
+
     virtual void close() override;
 
     bool FoundLinearizedHint() const

--- a/frmts/pdf/pdfobject.cpp
+++ b/frmts/pdf/pdfobject.cpp
@@ -979,13 +979,13 @@ class GDALPDFDictionaryPoppler : public GDALPDFDictionary
 class GDALPDFArrayPoppler : public GDALPDFArray
 {
   private:
-    Array *m_poArray;
+    const Array *m_poArray;
     std::vector<std::unique_ptr<GDALPDFObject>> m_v{};
 
     CPL_DISALLOW_COPY_ASSIGN(GDALPDFArrayPoppler)
 
   public:
-    GDALPDFArrayPoppler(Array *poArray) : m_poArray(poArray)
+    GDALPDFArrayPoppler(const Array *poArray) : m_poArray(poArray)
     {
     }
 
@@ -1334,7 +1334,7 @@ std::map<CPLString, GDALPDFObject *> &GDALPDFDictionaryPoppler::GetValues()
 /*                           GDALPDFCreateArray()                       */
 /************************************************************************/
 
-GDALPDFArray *GDALPDFCreateArray(Array *array)
+GDALPDFArray *GDALPDFCreateArray(const Array *array)
 {
     return new GDALPDFArrayPoppler(array);
 }

--- a/frmts/pdf/pdfobject.h
+++ b/frmts/pdf/pdfobject.h
@@ -404,7 +404,7 @@ class GDALPDFObjectPoppler : public GDALPDFObject
     virtual int GetRefGen() override;
 };
 
-GDALPDFArray *GDALPDFCreateArray(Array *array);
+GDALPDFArray *GDALPDFCreateArray(const Array *array);
 
 #endif  // HAVE_POPPLER
 

--- a/frmts/pdf/pdfsdk_headers_poppler.h
+++ b/frmts/pdf/pdfsdk_headers_poppler.h
@@ -51,10 +51,15 @@
 
 #include <Dict.h>
 
-#define private                                                                \
-    public /* Ugly! Catalog::optContent is private but we need it... */
+#if POPPLER_MAJOR_VERSION > 25 ||                                              \
+    (POPPLER_MAJOR_VERSION == 25 && POPPLER_MINOR_VERSION >= 2)
+#include <Catalog.h>
+#else
+/* Ugly! Catalog::optContent is private but we need it for ancient Poppler versions. */
+#define private public
 #include <Catalog.h>
 #undef private
+#endif
 
 #define private public /* Ugly! PDFDoc::str is private but we need it... */
 #include <PDFDoc.h>

